### PR TITLE
d3 based order book, all levels, animations, adjust liquidity function

### PR DESCRIPTION
**Pull request by imykolenko**
- [x] display all levels on linear scale
 - [x] d3 based animation
 - [x] adjust liquidity function to asymmetric gaussian + randomness
 - [x] dynamic label width to fit price length (int + required number of decimals)
 - [x] inverted market ticks calculation (non-linear because of 1 / price)
 
 ======
Screenshot 2024-06-18 at 1 15 56 PM

Screenshot 2024-06-18 at 1 15 47 PM

### CODE REVIEW

File: web-ui/src/components/Screens/HomeScreen/OrderBookWidget.tsx [hunk 1]
```diff
@@ -182,149 +137,346 @@ function directionStyle(direction: Direction) {
...
+        )
+        totalAmount = levels.reduce((acc, d) => acc + d.size, 0)
+      }
+
+      const tempText = svg
+        .select('.y-axis-mouse-projection')
+        .append('text')
+        .style('visibility', 'hidden')
+        .text(levelAmount.toFixed(tickDecimals + 1))
+      const maxLabelWidth = (tempText.node()?.getBBox().width ?? 35) * 0.95 + 7
```
Comment (bholzman) [inline web-ui/src/components/Screens/HomeScreen/OrderBookWidget.tsx:222]: is this perhaps a bit too magic? maybe at least a comment explaining the rationale?

**Review by bholzman — APPROVED** [review]
Looks great, nice job!

File: mocker/src/main/kotlin/co/chainring/mocker/core/Maker.kt [hunk 2]
```diff
@@ -292,22 +305,114 @@ class Maker(
...
+
+//    BTC/ETH
+//    val tickSize = BigDecimal("0.05")
+//    val initialPrice = BigDecimal(17.5)
+//    val marketToPeakStdDevFactor = 7.0
+
+//    BTC/USDC
+//    val tickSize = BigDecimal("1")
+//    val initialPrice = BigDecimal(68390.500)
+//    val marketToPeakStdDevFactor = 7.0
```
Comment (antonyshchenko) [inline mocker/src/main/kotlin/co/chainring/mocker/core/Maker.kt:402]: Did you mean to leave it commented out?

File: web-ui/src/components/Screens/HomeScreen/OrderBookWidget.tsx [hunk 3]
```diff
@@ -108,64 +47,156 @@ export function OrderBookWidget({
...
+            <div className="text-center">Empty</div>
+          ) : (
+            <OrderBookChart
+              tickDecimals={
+                side === 'Sell'
+                  ? (market.tickSize.decimalPlaces() === 0
+                      ? 5
+                      : market.tickSize.decimalPlaces()) +
+                    rawOrderBook.last.price.split('.')[0].length
+                  : market.tickSize.decimalPlaces()
```
Comment (antonyshchenko) [inline web-ui/src/components/Screens/HomeScreen/OrderBookWidget.tsx:186]: Nested ternary operators are a little hard to read. How about extracting this to a `memo`?

File: web-ui/src/components/Screens/HomeScreen/PricesWidget.tsx [hunk 4]
```diff
@@ -421,8 +421,8 @@ function OHLCChart({
       svg
         .select('.y-axis-current-price')
         .classed('hidden', !lastPrice)
-        .transition()
-        .duration(150)
+        // .transition()
+        // .duration(150)
```
Comment (antonyshchenko) [inline web-ui/src/components/Screens/HomeScreen/PricesWidget.tsx:425]: Did you mean to leave it commented out?

**Review by antonyshchenko — APPROVED** [review]

File: web-ui/src/index.css [hunk 5]
```diff
@@ -78,6 +78,16 @@
...
 
+
+.y-axis-order-book-last-price {
+  fill: #253447; /*darkBluishGray9*/
+}
+.y-axis-order-book-last-price text {
+  font-family: sans-serif;
+}
+
+.order-book{}
```
Comment (antonyshchenko) [inline web-ui/src/index.css:89]: Do we need this empty class?

Comment (imykolenko) [reply web-ui/src/index.css:89]: I use this class on svg to select with d3 specific group where content should be placed. Linter complains if there is no style definition.

File: web-ui/src/components/Screens/HomeScreen/PricesWidget.tsx (hunk 4 above)
Comment (imykolenko) [reply web-ui/src/components/Screens/HomeScreen/PricesWidget.tsx:425]: yes, as temporary measure. So far it looks like that blank screen problem is related to delayed animations catching up with effects that are caused by incoming data via web socket. I want to confirm on env, and then replace with proper solution.

File: mocker/src/main/kotlin/co/chainring/mocker/core/Maker.kt (hunk 2 above)
Comment (imykolenko) [reply mocker/src/main/kotlin/co/chainring/mocker/core/Maker.kt:402]: it is main function that just start swing window and draws price graph for debugging reasons. Updated to remove comments.

**Pull request merged**